luavm/src/lib/lua_values/mod.rs

```diff
@@ -0,0 +1,754 @@
+            LuaVal::from(self.to_int()? + other.to_int()?)
+        })
+    }
+
+    pub fn sub(&self, other: &LuaVal) -> Result<LuaVal, LuaError> {
+        Ok(if self.is_float() || other.is_float() {
+            LuaVal::from(self.to_float()? - other.to_float()?)
+        } else {
+            LuaVal::from(self.to_int()? - other.to_int()?)
+        })
+    }
+
+    pub fn mul(&self, other: &LuaVal) -> Result<LuaVal, LuaError> {
+        Ok(if self.is_float() || other.is_float() {
+            LuaVal::from(self.to_float()? * other.to_float()?)
+        } else {
+            LuaVal::from(self.to_int()? * other.to_int()?)
+        })
+    }
+
+    pub fn div(&self, other: &LuaVal) -> Result<LuaVal, LuaError> {
+        Ok(if self.is_float() || other.is_float() {
+            LuaVal::from(self.to_float()? / other.to_float()?)
+        } else {
+            LuaVal::from(self.to_int()? / other.to_int()?)
+        })
+    }
+
+    pub fn modulus(&self, other: &LuaVal) -> Result<LuaVal, LuaError> {
+        Ok(if self.is_float() || other.is_float() {
+            LuaVal::from(self.to_float()? % other.to_float()?)
+        } else {
+            LuaVal::from(self.to_int()? % other.to_int()?)
+        })
+    }
+
+    pub fn fdiv(&self, other: &LuaVal) -> Result<LuaVal, LuaError> {
+        Ok(if self.is_float() || other.is_float() {
+            LuaVal::from((self.to_float()? / other.to_float()?).floor())
+        } else {
+            LuaVal::from(self.to_int()? / other.to_int()?)
+        })
+    }
+
+    pub fn exp(&self, other: &LuaVal) -> Result<LuaVal, LuaError> {
+        Ok(LuaVal::from(self.to_float()?.powf(other.to_float()?)))
+    }
+}
+
+impl From<i64> for LuaVal {
+    /// Create an integer LuaVal.
+    fn from(int: i64) -> Self {
+        LuaVal {
+            encoded_ptr: to_raw_ptr(int) ^ LuaValKind::INT as usize,
+            owns: true,
+        }
+    }
+}
+
+impl From<f64> for LuaVal {
+    /// Create an integer LuaVal.
+    fn from(float: f64) -> Self {
+        LuaVal {
+            encoded_ptr: to_raw_ptr(float) ^ LuaValKind::FLOAT as usize,
+            owns: true,
+        }
+    }
+}
+
+impl From<String> for LuaVal {
+    /// Create a string LuaVal.
+    fn from(string: String) -> Self {
+        LuaVal {
+            encoded_ptr: to_raw_ptr(string) ^ LuaValKind::STR as usize,
+            owns: true,
+        }
+    }
+}
+
+impl From<LuaTable> for LuaVal {
+    /// Create an integer LuaVal.
+    fn from(table: LuaTable) -> Self {
+        LuaVal {
+            encoded_ptr: to_raw_ptr(Gc::new(table)) ^ LuaValKind::TABLE as usize,
+            owns: true,
+        }
+    }
+}
+
+impl Drop for LuaVal {
+    fn drop(&mut self) {
+        if self.owns {
+            unsafe {
+                match self.kind() {
+                    // NIL is a nullptr, so there is no need to free
+                    LuaValKind::NIL => (),
+                    LuaValKind::INT => {
+                        Box::from_raw(i64_ptr(self.encoded_ptr));
```

Comment:
I thought we were storing `INT` and so on as "raw" integers, which weren't malloc'd. Or maybe I'm wrong!

Reply:
I don't think I would be able to do pointer tagging in this case. If `encoded_ptr` were to be a raw integer, then I wouldn't be able to use any bits to check whether it is an int or not. This means that I would need an extra field (maybe `type`?) which tells me how to interpret the 64 bits. But if I have a pointer to something, I can store this information in the low-order bits.

Reply:
If you're not doing pointer tagging for ints, then there's no point in doing pointer tagging: currently you're getting all the worst of all worlds (extra complexity; extra branches) for none of the benefits.

Consider a 64 bit pointer. You use (say) 2 bits for tags. That leaves you 62 bits to do whatever you want. You can thus store a 62 bit in those 62 bits (possibly needing to left or right shift it, depending on where the tag bits are). What should you do if you have an int that takes 63 bits? That's where you have a second, boxed, version (called e.g. `LuaInt`) that seamlessly takes over.

Reply:
25% of the time, I won't be able to use those 62 bits to store an integer. If the integers are odd negative numbers (a quarter of all ints), that means I will have to box them into a `LuaInt`. Is there a performance gain from doing this, instead of simply boxing all the time? 

Doesn't this also mean that I will need to encode more types? (`LuaInt` vs raw int)
I can only encode 8 types (on 64 bit machines), and I also need to support `bool`s and `function`s. I currently have 5 types, and if I add the other 2 I will have a total of 7 types.

Reply:
There's nothing magical about 64 bits: you just have to choose an encoding. See e.g. https://github.com/softdevteam/packedvec/commit/98f42fd85de4cb974965d8abaa4d1dc27041051e

Reply:
I came up with [this](https://gist.github.com/rbartlensky/64339e5a1a3d4cbef99a8596471ea846). Is this closer to what you wanted originally? 

Most notable changes are in `LuaVal`, `clone`, and also in the `from` functions.

Reply:
The `ptr_type` field means that `LuaVal` takes up two machine words instead of one, which isn't what we want. We want to pack everything into one word.

Reply:
Ok, I see. But that means that my LuaValKind will have the following variants: Int, BoxedInt, Float, BoxedFloat, String, Nil, Table, right?

Reply:
What about something like [this](https://gist.github.com/rbartlensky/34d2c9023e20f4d5b1fafec6c4e4079f)?

Reply:
This looks more like it, except there shouldn't be an explicit `BOXED_INT` and so on: as I've before, when you have to box an int, you should have a struct called `LuaInt` (or `LuaBoxedInt` if you prefer) that's just "yet another" type of boxable thing.

Reply:
I *think* [this](https://gist.github.com/rbartlensky/60fff14d6652b909fb539b404c448108) might be closer to what you are suggesting. Please note that a `Box<LuaObj>` has a size of 16 bytes, therefore I need to wrap it into another box, and store that address in `LuaVal`.

Reply:
At first glance, this looks more like it!